drift-client/src/main/java/io/airlift/drift/client/DriftMethodInvocation.java

```diff
@@ -226,7 +245,19 @@ else if (exceptionClassification.getHostStatus() == OVERLOADED) {
 
             // A request to down or overloaded server is not counted as an attempt, and retries are not delayed
             if (exceptionClassification.getHostStatus() != NORMAL) {
-                nextAttempt();
+                Optional<A> selectedAddress = addressSelector.selectAddress(addressSelectionContext, attemptedAddresses.keySet());
```

Comment:
I'm not sure I like this code organization.  Maybe instead of passing the address to `nextAttempt`, we let next attempt select the address and update stats, as it is doing now, and then check if we need to delay for that address.  The obvious down side is we could double delay from request failure and connection reattempt.  I think for that we could simply pass a boolean `noConnectDelay` to `nextAttempt` to signal that we have already delayed the request.  I'm hoping this will keep the code fairly simple to follow.

Reply:
Also, I think we should consider logging at debug when delaying a connect attempt.

Reply:
> I'm not sure I like this code organization. Maybe instead of passing the address to nextAttempt, we let next attempt select the address and update stats, as it is doing now, and then check if we need to delay for that address. The obvious down side is we could double delay from request failure and connection reattempt. I think for that we could simply pass a boolean noConnectDelay to nextAttempt to signal that we have already delayed the request. I'm hoping this will keep the code fairly simple to follow.

I reorganized the code. I added a separate map, that tracks only connection failed attempts. Also i don't check if the request is not delayed twice (based on global retries and connection attempt retries).